File: skills/cumcm-code-writer/references/algorithm_library/02_评价决策/entropy_weight/python/main.py
"""熵权法 · 客观赋权（Python 最小可运行模板）

步骤：正/负向指标极差标准化（补 0.002 避免零值）→ 计算比重 → 熵值 → 差异系数归一化得权重。
接口：solve(X, positive) -> 权重；运行：python main.py
"""
from __future__ import annotations

import sys
from pathlib import Path

import numpy as np

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "_common" / "python"))
from mmlib import demo, report, save_table  # noqa: E402


def build_demo() -> dict:
    X = np.array([[92, 0.83, 3.1, 8], [78, 0.71, 4.5, 6], [85, 0.90, 2.7, 9],
                  [70, 0.64, 5.2, 5], [88, 0.79, 3.6, 7]], dtype=float)
    return {"X": X, "positive": [True, True, False, True]}


def normalize(X, positive):
    X = np.asarray(X, float)
    Z = np.empty_like(X)
    for j in range(X.shape[1]):
        col = X[:, j]
        span = col.max() - col.min()
        Z[:, j] = (col - col.min()) / span if positive[j] else (col.max() - col) / span
    return Z + 0.002


def solve(X, positive):
    Z = normalize(X, positive)
    P = Z / Z.sum(axis=0, keepdims=True)
    n = Z.shape[0]
    E = -(P * np.log(P)).sum(axis=0) / np.log(n)
    d = 1 - E
    return d / d.sum(), E


if __name__ == "__main__":
    data = demo(__file__, build_demo)
    weight, entropy = solve(data["X"], data["positive"])
    report("熵权法", 权重=np.round(weight, 4).tolist(), 熵值=np.round(entropy, 4).tolist())
    save_table(__file__, "指标权重", {"指标": [f"X{i + 1}" for i in range(len(weight))],
                                      "熵值": np.round(entropy, 6), "权重": np.round(weight, 6)})

File: skills/cumcm-code-writer/references/algorithm_library/02_评价决策/entropy_weight/matlab/main.m
% 熵权法 · 客观赋权（MATLAB 最小可运行模板）
here = fileparts(mfilename('fullpath'));
addpath(fullfile(here, '..', '..', '..', '_common', 'matlab'));

X = [92 0.83 3.1 8; 78 0.71 4.5 6; 85 0.90 2.7 9; 70 0.64 5.2 5; 88 0.79 3.6 7];
positive = [true true false true];

Z = zeros(size(X));
for j = 1:size(X, 2)
    col = X(:, j);
    if positive(j)
        Z(:, j) = (col - min(col)) / (max(col) - min(col));
    else
        Z(:, j) = (max(col) - col) / (max(col) - min(col));
    end
end
Z = Z + 0.002;
P = Z ./ sum(Z, 1);
n = size(Z, 1);
E = -sum(P .* log(P), 1) / log(n);
d = 1 - E;
w = d / sum(d);

mmlib_report('熵权法', sprintf('权重=%s', mat2str(round(w, 4))), sprintf('熵值=%s', mat2str(round(E, 4))));
T = table("X" + string(1:numel(w))', E', w', 'VariableNames', {'指标', '熵值', '权重'});
mmlib_savetable(here, 'entropy_weight', T);
